### PR TITLE
Added x11 32bit and Windows 32bit for compiling

### PR DESCRIPTION
**Pull request by MasterPolska123**
So i needed to build this plugin for my 32bit Linux and also I made Windows 32bit build. As i saw in the terminal output there are no errors. Also 32bit linux build works perfectly on my 32bit laptop.

### CODE REVIEW

**Comment by MasterPolska123** [comment]
I did PR cause my work may help other people so I just decided to make pull request on this repository. Hope You'll enjoy my work

File: Dockerfile.win64 [hunk 1]
```diff
@@ -18,3 +18,9 @@ RUN scons -c platform=windows
 
 RUN scons platform=windows prefix="${FINAL_TARGET_DIR}"
 
+
+WORKDIR /opt/godot-videodecoder/ffmpeg-static
+RUN ./build.sh -B -p windows -T "$THIRDPARTY_DIR/win32"
+WORKDIR /opt/godot-videodecoder
+RUN scons -c platform=win32
+RUN scons platform=win32 prefix="${FINAL_TARGET_DIR}"
```
Comment (jamie-pate) [inline Dockerfile.win64:26]: This looks like it should have been removed?

File: build_gdnative.sh [hunk 2]
```diff
@@ -67,13 +66,22 @@ id=$(docker create godot-videodecoder-x11)
...
+docker cp $id:/opt/target/x11_32 $ADDON_BIN_DIR/
+docker rm -v $id
+
 echo "extracting $ADDON_BIN_DIR/osx"
 id=$(docker create godot-videodecoder-osx)
 docker cp $id:/opt/target/osx $ADDON_BIN_DIR/
 docker rm -v $id
 
-echo "extracting $ADDON_BIN_DIR/wi64"
+echo "extracting $ADDON_BIN_DIR/win64"
```
Comment (jamie-pate) [inline build_gdnative.sh:79]: :flushed:

**Review by jamie-pate — CHANGES_REQUESTED** [review]

File: Dockerfile.win64 (hunk 1 above)
Comment (MasterPolska123) [reply Dockerfile.win64:26]: Yeah, i forgot to remove it :/ I'm fixing this right now.

Comment (MasterPolska123) [reply Dockerfile.win64]: Ok. Should be ok. If you don't see the changes see master branch of my fork. I don't know that i fixed it that way, you can see it in this PR. I'm very new to Github's PRs

**Review by MasterPolska123 — COMMENTED** [review]
Now should be ok. If you don't see changes take a look at master branch of my fork of this repository

**Comment by MasterPolska123** [comment]
And i saw now i forgot to include SConstruct file prepared for those two platforms. I'll include it here as soon as possible or in a new pr. Really dunno why did I make such a mistake. sorry about that

File: Dockerfile.win64 [hunk 3]
```diff
@@ -16,5 +16,4 @@ WORKDIR /opt/godot-videodecoder
 COPY . .
 RUN scons -c platform=windows
 
-RUN scons platform=windows prefix="${FINAL_TARGET_DIR}"
-
+RUN scons platform=windows prefix="${FINAL_TARGET_DIR}
```
Comment (jamie-pate) [inline Dockerfile.win64:19]: Missing end quote, could just use `git checkout origin/master Docker.win64`

**Review by jamie-pate — CHANGES_REQUESTED** [review]
Also I think the scons changes you mentioned should be on this pr

**Review by MasterPolska123 — COMMENTED** [review]
ok. end quote fixed. also i guess that github added modified SConstruct file to the pull request so everything should be fine when there would be no issues with SConstruct file :)

File: SConstruct [hunk 4]
```diff
@@ -35,18 +35,23 @@ if env['platform'] == 'x11':
...
 tool_prefix = ''
 if os.name == 'posix' and env['platform'] == 'win64':
     tool_prefix = "x86_64-w64-mingw32-"
-    if (os.getenv("MINGW64_PREFIX")):
-        tool_prefix = os.getenv("MINGW64_PREFIX")
     env['SHLIBSUFFIX'] = '.dll'
     env.Append(CPPDEFINES='WIN32')
-
+if os.name == 'posix' and env['platform'] == 'win32':
+    tool_prefix = "i686-w64-mingw32-"
```
Comment (jamie-pate) [inline SConstruct:52]: I think this should be w32 not w64?

File: SConstruct [hunk 5]
```diff
@@ -56,7 +61,9 @@ if os.name == 'posix' and env['platform'] == 'osx':
 globs = {
     'x11': '*.so.[0-9]*',
     'win64': '../bin/*-[0-9]*.dll',
-    'osx': '*.[0-9]*.dylib'
+    'osx': '*.[0-9]*.dylib',
+	 'x11_32': '*.so.[0-9]*',
```
Comment (jamie-pate) [inline SConstruct:65]: Mixed tabs and spaces or just the wrong number of spaces?

Should match indentation either way

File: SConstruct (hunk 4 above)
Comment (jamie-pate) [reply SConstruct:52]: I think this should be w32 not w64?

**Review by jamie-pate — CHANGES_REQUESTED** [review]

File: SConstruct (hunk 4 above)
Comment (MasterPolska123) [reply SConstruct:52]: erm... On my pc there's installed an `i686-w64-mingw32` package so i wrote `i686-w64-mingw32-` there in `tool_prefix`. I also didn't find an linux package named: `i686-w32-mingw32`. Also `i686-w64-mingw32` works with compilation for 32bit platforms for me so i don't think that there's need to change `i686-w64-mingw32` to `i686-w32-mingw32`

File: SConstruct (hunk 5 above)
Comment (MasterPolska123) [reply SConstruct:65]: Wrong number of spaces that my code editor didn't show me while coding SConstruct file for 32bit platforms compilation

**Review by MasterPolska123 — COMMENTED** [review]
fixed requested changes

File: SConstruct (hunk 4 above)
Comment (jamie-pate) [reply SConstruct:52]: Hmm, confusing

**Review by jamie-pate — APPROVED** [review]

**Comment by jamie-pate** [comment]
Hey, did you have any changes to ffmpeg-static to back this?
I'm getting some issues compiling for the x11_32 and win32 platforms:

```
*** Building FFmpeg ***
Makefile:2: ffbuild/config.mak: No such file or directory
Makefile:40: /tools/Makefile: No such file or directory
Makefile:41: /ffbuild/common.mak: No such file or directory
Makefile:90: /libavutil/Makefile: No such file or directory
Makefile:90: /ffbuild/library.mak: No such file or directory
Makefile:92: /fftools/Makefile: No such file or directory
Makefile:93: /doc/Makefile: No such file or directory
Makefile:94: /doc/examples/Makefile: No such file or directory
Makefile:159: /tests/Makefile: No such file or directory
make: *** No rule to make target '/tests/Makefile'.  Stop.
+ EXTRA_LIBS= -lpthread -lm 
+ PKG_CONFIG_PATH=/opt/godot-videodecoder/ffmpeg-static/target/lib/pkgconfig ./configure --prefix=/opt/godot-videodecoder/thirdparty/win32 --pkg-config-flags=--static --extra-cflags=-I/opt/godot-videodecoder/ffmpeg-static/target/include --extra-ldflags=-L/opt/godot-videodecoder/ffmpeg-static/target/lib --extra-libs= -lpthread -lm  --bindir=/opt/godot-videodecoder/ffmpeg-static/bin --disable-everything --disable-debug --disable-gpl --disable-nonfree --disable-programs --enable-shared --disable-static --enable-decoder=libopus --enable-decoder=opus --enable-decoder=vp9 --enable-decoder=libvorbis --enable-decoder=vorbis --enable-parser=vp9 --enable-parser=opus --enable-parser=vorbis --enable-demuxer=matroska --enable-demuxer=opus --enable-demuxer=vorbis --enable-libopus --enable-libvorbis --enable-opengl --disable-vaapi --disable-vdpau
ERROR: opengl not found.```

```
Install file: "thirdparty/x11_32/lib/libavdevice.so.58" as "/opt/target/x11_32/libavdevice.so.58"
gcc -o src/gdnative_videodecoder.os -c -std=gnu11 -fPIC -g -Ithirdparty/x11_32/include -Igodot_include src/gdnative_videodecoder.c
src/gdnative_videodecoder.c: In function 'godot_videodecoder_get_supported_ext':
src/gdnative_videodecoder.c:511:9: warning: return from incompatible pointer type [-Wincompatible-pointer-types]
  return supported_ext;
         ^
scons: building terminated because of errors.
scons: *** [bin/x11_32/libgdnative_videodecoder.so] Implicit dependency `/usr/lib/i386-linux-gnu/libc_nonshared.a' not found, needed by target `bin/x11_32/libgdnative_videodecoder.so'.
The command '/bin/sh -c scons platform=x11_32 prefix="${FINAL_TARGET_DIR}"' returned a non-zero code: 2```

**Comment by jamie-pate** [comment]
Can you try this build and see if it works for win 32 and x11_32?

**Comment by JezSonic** [comment]
Sure, of course. Just give me a minute. (I'm a student. I have lessons now)

**Comment by JezSonic** [comment]
x11_32 build from build from GitHub Actions is not working. Error says
```
**ERROR: open_dynamic_library: Can't open dynamic library: /home/godot/Pulpit/Foxy-Adventure/addons/bin/x11_32/libgdnative_videodecoder.so. Error: /home/godot/Pulpit/Foxy-Adventure/addons/bin/x11_32/libgdnative_videodecoder.so: wrong ELF class: ELFCLASS64**
```
Maybe you should configure actions to do the compilation on 32bit OS?

**Comment by jamie-pate** [comment]
It was a long shot that it would work the first time.. just gave to get the cross compiler set up properly... Getting closer though

**Comment by jamie-pate** [comment]
yup 
target/x11_32/libgdnative_videodecoder.so: ELF 64-bit LSB shared object

:(

**Comment by JezSonic** [comment]
I will send you my compilation for x11_32 when I'll be at home. Also. I tried to compile that plugin for android and i used precompiled ffmpeg downloaded from web. While compilation ndk prints lots of undefined references in src/gdnative_videodecoder.c and terminates the build while everything is ok in jni/Application.mk and jni/Android.mk. Any thoughts how to fix it?

**Comment by jamie-pate** [comment]
1) the idea is to have automated builds, no point uploading builds as they will just be out of date soon

2) no idea, undefined references are probably the functions that it expects to link against from the ffmpeg library? iOS and Android platforms are very different from Windows

**Comment by JezSonic** [comment]
Yeah, but libraries and those files are linked properly in jni/Android.mk file

**Comment by jamie-pate** [comment]
Probably have to put in the code for the get_ticks_msec() etc from the Android platform?

It would help more to have the complete output

**Comment by JezSonic** [comment]
I did something and now I have 30 kilobytes output files from Android ndk. I'm not sure that I linked libs properly in Android.mk file

**Comment by jamie-pate** [comment]
https://github.com/jamie-pate/godot-videodecoder/releases/tag/0.0.1-draft.14 this one has all the right file types, but need help testing for x11_32 and win32 :D

**Comment by JezSonic** [comment]
I'll test it as soon as possible. Also i was talking with my friend about my issue with compiling videodecoder for android and he said it may be "cause of different preprocessor on android". I asked him to help me and he agreed so today we're gonna work with it. Also can you add me on discord so i would ask you on discord if i have questions about godot videodecoder for android (add me anyway, haha - just kidding)?

**Comment by JezSonic** [comment]
so x11_32 from latest auto-builded release works, win32 too. I also saw in debug logs, that win32 requires this file to run `libgcc_s_sjlj-1.dll`

**Pull request merged**